CYCLONEDDS_QUALITY_DECLARATION.md

```diff
@@ -0,0 +1,150 @@
+This document is a declaration of software quality for the `CycloneDDS` external dependency imported by the rmw\_cyclonedds package, based on the guidelines in [REP-2004](https://www.ros.org/reps/rep-2004.html).
+
+# `CycloneDDS` Quality Declaration
+
+The package `CycloneDDS` claims to be in the **Quality Level 4** category.
+
+The `CycloneDDS` project meets a substantial number of the criteria stated for software quality.
+The developers have taken many steps to ensure that `CycloneDDS` is a software product which can be relied upon, however there is a significant lack of documentation regarding the interpretation of versions, features, and testing.
+Even minimal investments in these areas would contribute to the quality of the project, and could increase the level category for `CycloneDDS` in a subsequent review.
+
+Below are the rationales, notes, and caveats for this claim, organized by each requirement listed in the [Package Requirements for Quality Level 4 in REP-2004](https://www.ros.org/reps/rep-2004.html).
+
+## Version Policy [1]
+
+### Version Scheme [1.i]
+
+There is a three-part version attached to `CycloneDDS` releases.
+The sources define the version parts as `MAJOR`.`MINOR`.`PATCH`.
+The CMake sources configure the compatibility mode to [`SameMajorVersion`](https://cmake.org/cmake/help/latest/module/CMakePackageConfigHelpers.html#generating-a-package-version-file), indicating that breaking changes are accompanied by a change to the `MAJOR` version part.
+There is no explicit policy regarding versioning in the documentation for `CycloneDDS`, however there is some information in the [Releases section](https://www.eclipse.org/projects/handbook/#release) of the Eclipse project handbook which discusses _Major_, _Minor_, and _Service_ release criteria.
+It is possible that these release categories align align with changes to the three version parts, but this is not explicitly stated in either `CycloneDDS` documentation or in the Eclipse project handbook.
+
+### Version Stability [1.ii]
+
+The latest release of `CycloneDDS`, `0.5.1`, does not have a stable version, i.e. `>= 1.0.0`.
```

Comment:
Since exact latest versions change frequently, I think it's enough to just say it's not yet above 1.0. Or if you want to add it, maybe point to where someone can find the latest version?

Reply:
0.5.x is a stable release.

Reply:
@rotu is right, don’t be fooled by the 0 at the beginning (0.1 was already a stable release for that matter). There are various reasons for it still being 0.x, but those have to do with functionality, websites and documentation, process. Quality and stability having nothing to do with it.

I think it’d be silly to document here what the latest released version of Cyclone DDS is, but we can keep it updated.

Reply:
This is great news, really, but points to a lack of documentation. For packages that follow semver, `1.0.0` is the level that marks stability. As discussed in [1.i], the versioning scheme for `CycloneDDS` isn't well-documented, so I shouldn't have jumped to enforcing semver's stability requirements here.

At the risk of getting a bit pedantic, I believe that this section specifically describes version numbers in relation to stability. I'm glad that the product is stable, but I believe the intent here is to be able to determine stability of a product based on the version alone. This is especially important for projects that alternate between development and production versions like early Linux kernels.

If all versions `>= 0.1.0` are stable, is that documented somewhere?

Reply:
It follows semantic versioning with one twist: major version 0 is stable as well. It is simply version 0.x because I followed the Eclipse Incubation-stage versioning convention when I was forced to pick a number. I am not going to arbitrarily bump it.

There is much better proof than a statement in documentation and that’s its track record: an almost complete absence of regressions and careful consideration of backwards compatibility extending all the way back to the initial contribution. I get that that is a bit unusual in this brave new world of depending on totally unstable software.

Reply:
If that is the policy of CycloneDDS, then that's perfect for this QD. The goal definitely isn't to force one particular versioning convention on anyone, as long as consumers of the package know what to expect. All @cottsay is looking for was a place to point to so that if someone else comes along, reads the QD and and asks why a 0.x release is considered stable they can verify the justification. A change log is crucial in providing evidence that a policy is followed, but it's not a declaration of a policy itself.

Reply:
https://github.com/eclipse-cyclonedds/cyclonedds/pull/507